### PR TITLE
performance fix for toInt in Cypher

### PR DESCRIPTION
**Pull request by jexp**

### CODE REVIEW

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/ToIntFunction.scala [hunk 1]
```diff
@@ -34,14 +34,26 @@ case class ToIntFunction(a: Expression) extends NullInNullOutExpression(a) {
...
   def compute(value: Any, m: ExecutionContext)(implicit state: QueryState): Any = a(m) match {
     case v: Number =>
       v.longValue()
     case v: String =>
       try {
-        val d = BigDecimal(v)
-        if (d <= Long.MaxValue && d >= Long.MinValue) d.toLong
-        else throw new CypherTypeException(s"integer, $v, is too large")
+        val longValue = if (v.length < 20) parseLong(v) else null
+        if (longValue != null) longValue
```
Comment (chrisvest) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/ToIntFunction.scala]: Why not collapse these two conditionals?

**Comment by lutovich** [comment]
Imo we can avoid `BigDecimal` conversion and magic `20` all together. Why not just optimistically call `d.toLong` and examine string value if it throws?

**Comment by jexp** [comment]
@lutovich yes this is my latest variant

```
  def compute(value: Any, m: ExecutionContext)(implicit state: QueryState): Any = a(m) match {
    case v: Number =>
      v.longValue()
    case v: String =>
      try {
        java.lang.Long.parseLong(value)
      } catch {
        try {
          val d = BigDecimal(v)
          if (d <= Long.MaxValue && d >= Long.MinValue) d.toLong
          else throw new CypherTypeException(s"integer, $v, is too large")
        } catch {
          case e: NumberFormatException =>
            null
        }
      }
    case v =>
      throw new ParameterWrongTypeException("Expected a String or Number, got: " + v.toString)
  }
```

We still need the 2 try-catch blocks, as they have different semantics.

**Comment by jexp** [comment]
Just looked at `Long.parseLong` which is also an awful method. I hope they have an intrinsic for it or a native call.

```
public static long parseLong(String s, int radix)
              throws NumberFormatException
    {
        if (s == null) {
            throw new NumberFormatException("null");
        }

        if (radix < Character.MIN_RADIX) {
            throw new NumberFormatException("radix " + radix +
                                            " less than Character.MIN_RADIX");
        }
        if (radix > Character.MAX_RADIX) {
            throw new NumberFormatException("radix " + radix +
                                            " greater than Character.MAX_RADIX");
        }

        long result = 0;
        boolean negative = false;
        int i = 0, len = s.length();
        long limit = -Long.MAX_VALUE;
        long multmin;
        int digit;

        if (len > 0) {
            char firstChar = s.charAt(0);
            if (firstChar < '0') { // Possible leading "+" or "-"
                if (firstChar == '-') {
                    negative = true;
                    limit = Long.MIN_VALUE;
                } else if (firstChar != '+')
                    throw NumberFormatException.forInputString(s);

                if (len == 1) // Cannot have lone "+" or "-"
                    throw NumberFormatException.forInputString(s);
                i++;
            }
            multmin = limit / radix;
            while (i < len) {
                // Accumulating negatively avoids surprises near MAX_VALUE
                digit = Character.digit(s.charAt(i++),radix);
                if (digit < 0) {
                    throw NumberFormatException.forInputString(s);
                }
                if (result < multmin) {
                    throw NumberFormatException.forInputString(s);
                }
                result *= radix;
                if (result < limit + digit) {
                    throw NumberFormatException.forInputString(s);
                }
                result -= digit;
            }
        } else {
            throw NumberFormatException.forInputString(s);
        }
        return negative ? result : -result;
    }
```

**Comment by jexp** [comment]
Weird that there is no intrinsic for parse methods.

This one is up to 20% faster than Long.parseLong, but I was surprised that even that ugly method (Long.parseLong) with 272 instructions is inlined.

``` java
import org.junit.Test;

import java.util.concurrent.TimeUnit;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

/**
 * @author mh
 * @since 29.05.16
 */
public class ParseLongTest {

    public static long parseLong(String s) {
        int len = s.length();
        if (len != 0 && len < 21) {
            int i = 0, p = 1;
            int c = s.charAt(0);
            if (c == '-') { p = -1; i = 1; }
// optional
//            else if (c == '+') { i = 1; }
            if (i < len) {
                long result = 0;
                for (; i < len && result > -1; i++) {
                    c = s.charAt(i) - '0';
                    result = (c > -1 && c < 10) ? (result << 3) + (result << 1) + c : -1;
                }
                if (i == len && result > -1) return p * result;
                if (result == Long.MIN_VALUE && p == -1) return result;
            }
        }
        throw new NumberFormatException(s);
    }

    public static final int TIME_LIMIT = 35000;
    public static final int WARMUP_RANGE = 100;
    public static final int RANGE = 10;

    @Test
    public void testParseLong() throws Exception {
        testRange(WARMUP_RANGE);
        testRange(WARMUP_RANGE);
        long time = System.nanoTime();
        testRange(RANGE);
        time = System.nanoTime() - time;
        time = TimeUnit.NANOSECONDS.toMillis(time);
        System.out.println("time for 2/" + RANGE + "th of the INT value space (" + (Integer.MAX_VALUE / RANGE - Integer.MIN_VALUE / RANGE) + "): " + time + " ms.");
        assertEquals("time limit " + TIME_LIMIT, true, time < TIME_LIMIT);
    }

    @Test
    public void testParseLongOriginal() throws Exception {
        testRangeLong(WARMUP_RANGE);
        testRangeLong(WARMUP_RANGE);
        long time = System.nanoTime();
        testRangeLong(RANGE);
        time = System.nanoTime() - time;
        time = TimeUnit.NANOSECONDS.toMillis(time);
        System.out.println("time for 2/" + RANGE + "th of the INT value space (" + (Integer.MAX_VALUE / RANGE - Integer.MIN_VALUE / RANGE) + "): " + time + " ms.");
        assertEquals("time limit " + TIME_LIMIT, true, time < TIME_LIMIT);
    }

    private void testRangeLong(int range) {
        for (long l = Integer.MIN_VALUE / range; l < Integer.MAX_VALUE / range; l++) {
            assertEquals(l, Long.parseLong(String.valueOf(l)));
        }
        assertEquals((long) Integer.MAX_VALUE, Long.parseLong(String.valueOf(Integer.MAX_VALUE)));
        assertEquals(Long.MAX_VALUE, Long.parseLong(String.valueOf(Long.MAX_VALUE)));
        assertEquals((long) Integer.MIN_VALUE, Long.parseLong(String.valueOf(Integer.MIN_VALUE)));
        assertEquals(Long.MIN_VALUE, Long.parseLong(String.valueOf(Long.MIN_VALUE)));
    }

    private void testRange(int range) {
        for (long l = Integer.MIN_VALUE / range; l < Integer.MAX_VALUE / range; l++) {
            assertEquals(l, parseLong(String.valueOf(l)));
        }
        assertEquals((long) Integer.MAX_VALUE, parseLong(String.valueOf(Integer.MAX_VALUE)));
        assertEquals(Long.MAX_VALUE, parseLong(String.valueOf(Long.MAX_VALUE)));
        assertEquals((long) Integer.MIN_VALUE, parseLong(String.valueOf(Integer.MIN_VALUE)));
        assertEquals(Long.MIN_VALUE, parseLong(String.valueOf(Long.MIN_VALUE)));
    }

    @Test
    public void testLongMinValue() throws Exception {
        assertEquals(Long.MIN_VALUE, parseLong(String.valueOf(Long.MIN_VALUE)));
    }

    @Test
    public void testErrors() throws Exception {
        String[] invalid = {"", "a", "123" + String.valueOf((char) ('0' - 1)), 123 + String.valueOf((char) ('9' + 1)),
                String.valueOf((char) ('0' - 1)), String.valueOf((char) ('9' + 1)), "abc", "\0", "\n", "\t", "\b",
                "- 1", "+10", "-", "+", "3874892374092743920743289748932740", "-3874892374092743920743289748932740",
                String.valueOf(Long.MAX_VALUE) + "0", String.valueOf(Long.MIN_VALUE) + "0", "9223372036854775808", "-9223372036854775809"};
        for (String s : invalid) {
            try {
//                System.out.println("s = \"" + s + "\"");
                parseLong(s);
                fail("Didn't fail \"" + s + "\"");
            } catch (NumberFormatException nfe) { /*ignore*/ }
        }
    }
}
```

**Comment by systay** [comment]
This PR is failing on the build server, @jexp

**Comment by systay** [comment]
This PR is not compiling at the moment

**Comment by systay** [comment]
After it has been fixed, could you share profiling that shows that this change fixes the situation?

**Comment by pontusmelke** [comment]
~~@jexp You don't seem to have pushed the code that uses `longValue`?~~

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/ToIntFunction.scala [hunk 2]
```diff
@@ -39,12 +39,17 @@ case class ToIntFunction(a: Expression) extends NullInNullOutExpression(a) {
...
-        val d = BigDecimal(v)
-        if (d <= Long.MaxValue && d >= Long.MinValue) d.toLong
-        else throw new CypherTypeException(s"integer, $v, is too large")
+        java.lang.Long.parseLong(v)
       } catch {
-        case e: NumberFormatException =>
-          null
+        case e: Exception =>
+        try {
+          val d = BigDecimal(v)
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/ToIntFunction.scala:46]: For what values `v` will `BigDecimal(v)` + range check succeed for which `parseLong` will fail?

Comment (jexp) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/ToIntFunction.scala:46]: 1E10

**Comment by jexp** [comment]
@pontusmelke you mean my own implementation? I wanted some feedback on it first :)

**Comment by pontusmelke** [comment]
@jexp no sorry I misread

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/ToIntFunction.scala (hunk 2 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/ToIntFunction.scala:46]: Ah, got it.

**Comment by jexp** [comment]
@pontusmelke do you want me to push the code for my parseLong ? it makes quite a bit of a difference.

In general it's scary to see that parsing 10M strings into numbers takes 4-6 seconds in Java.

**Pull request merged**